### PR TITLE
Feature/qc step ds kraken

### PR DESCRIPTION
**Pull request by nickp60**
This adds downsampling to 20k reads, kraken, and merging into a krona report.  This also involved modifying the downsampling rule to optionally output files in cases where the desired depth exceeds the input depth; the default and previous behavior was to error.

Kraken variables `report` were renamed to `kreport` to avoid a snakemake error due to `report` being a snakemake keyword

### CODE REVIEW

**Comment by nickp60** [comment]
The following test manifest was used:
```
experiment.identifier   file_type       file_path       notes
473_A   fq1     /home/watersn/GitHub/vdblab-shotgun/tmppre_medium/hostdepleted/473_R1.fastq.gz  none
473_A   fq2     /home/watersn/GitHub/vdblab-shotgun/tmppre_medium/hostdepleted/473_R2.fastq.gz  none
473_A   host_fq1        /home/watersn/GitHub/vdblab-shotgun/tmppre_medium/hostdepleted/473_all_host_reads_R1.fastq.gz   none
473_A   host_fq2        /home/watersn/GitHub/vdblab-shotgun/tmppre_medium/hostdepleted/473_all_host_reads_R2.fastq.gz   none
473_A   sortmerna_report        /home/watersn/GitHub/vdblab-shotgun/tmppre_medium/reports/473_sortmerna.stats   none
473_B   fq1     /home/watersn/GitHub/vdblab-shotgun/tmppre_tiny/hostdepleted/473_R1.fastq.gz    none
473_B   fq2     /home/watersn/GitHub/vdblab-shotgun/tmppre_tiny/hostdepleted/473_R2.fastq.gz    none
473_B   host_fq1        /home/watersn/GitHub/vdblab-shotgun/tmppre_tiny/hostdepleted/473_all_host_reads_R1.fastq.gz     none
473_B   host_fq2        /home/watersn/GitHub/vdblab-shotgun/tmppre_tiny/hostdepleted/473_all_host_reads_R2.fastq.gz     none
473_B   sortmerna_report        /home/watersn/GitHub/vdblab-shotgun/tmppre_tiny/reports/473_sortmerna.stats     none
```

**Comment by miraep8** [comment]
👍 👍 Thanks for putting this together! Overall I think it looks pretty good (I will try to go through and do a more detailed review)! I did run into this (potential user error) when I tried this out though (isabl analyses `52938` and `52937` for example). Tracking it down a bit it seems like the error is occurring at the `downsample_fastq` stage from what seems to be an issue with the `input` to that rule returning blank, so I suspect that might be the source of it?

It also made me want to ask a bit about the downsample step: in the case that we have multiple fastqs for each R1, would we want to run this downsample and subsequent steps on each of the fastqs separately (I think this might be my vote - as it would enable us to potentially look for outliers amongst the experiments) or on just the merged fastq (which I guess in practice would just be the first of the experiment fastqs, and which would still be useful!)? Either way works, but I was a bit confused by the current set up.

**Comment by miraep8** [comment]
Looking at it more though I think my follow up question might be coming from (probably my) confusion of `sample` versus `experiment` here. I will probably need to look at it in more detail to figure it out what's actually going on here, so sorry if my question doesn't make sense! 😅

**Comment by nickp60** [comment]
Sorry for the confusion! I was re-using wildcards but just pushed a change where the relevant ones are called `exp` for experiment rather than sample. And yes -- when we have multiple fastq pairs (eg multiple flowcells, lanes have already been merged at this point), this workflow should run downsampling per-fastq pair, kraken  per downsampled fastq pair, and kraken2krona per downsampled fastq pair.  Both the beta diversity and the krona rules aggregated the experiments to a single sample-level result file.

Can you send a log file of the case where the input is empty?

**Comment by nickp60** [comment]
And I agree that its a bit of a departure from the rest of the workflow.  Not sure if there is a better place for it, or if we just need extra documentation on how it differs.

File: workflow/rules/qc.smk [hunk 1]
```diff
@@ -84,17 +99,27 @@ localrules:
 
 def get_all_outputs(wildcards):
     return [
-        f"qc/{config['sample']}_qc_merged_R1.fq",
-        f"qc/{config['sample']}_qc_merged_R2.fq",
-        f"qc/{config['sample']}_qc_stats.txt"
+        f"qc/{config['sample']}_qc_merged_R1.fq.gz",
+        f"qc/{config['sample']}_qc_merged_R2.fq.gz",
```
Comment (miraep8) [inline workflow/rules/qc.smk:103]: Nice, thanks for catching/fixing this!

**Review by miraep8 — APPROVED** [review]
Other than the isabl thing this looks good, happy for this to be merged once that is resolved!

**Pull request merged**